File: src/lib/models/personality.ts
export const personalityAxisSpectrum: Record<PersonalityAxis, { low: string; high: string }> = {
    Assertiveness: {
        low: 'Accommodating',
        high: 'Dominant'
    },
    StructurePreference: {
        low: 'Spontaneous',
        high: 'Methodical'
    },
    Openness: {
        low: 'Practical',
        high: 'Idealistic'
    },
    Sociability: {
        low: 'Reserved',
        high: 'Outgoing'
    },
    Influence: {
        low: 'Suggestive',
        high: 'Persuasive'
    }
};
export type PersonalitySnapshot = {
    assertiveness: number;
    structure_preference: number;
    openness: number;
    sociability: number;
    influence: number;

    assertiveness_description: string;
    structure_preference_description: string;
    openness_description: string;
    sociability_description: string;
    influence_description: string;
};
export type PersonalityAxis =
    | 'Assertiveness'
    | 'StructurePreference'
    | 'Openness'
    | 'Sociability'
    | 'Influence';
export const allPersonalityAxes: PersonalityAxis[] = [
    'Assertiveness',
    'StructurePreference',
    'Openness',
    'Sociability',
    'Influence'
];
export type PersonalityAxisValue = {
    value: number;
    desc: string;
};
export const personalityAxisIcons: Record<PersonalityAxis, string> = {
    Assertiveness: '🗣️',
    StructurePreference: '📅',
    Openness: '💡',
    Sociability: '👥',
    Influence: '🎯'
};
export const personalityAxisLabels: Record<PersonalityAxis, string> = {
    Assertiveness: 'Assertiveness',
    StructurePreference: 'Structure Preference',
    Openness: 'Openness',
    Sociability: 'Sociability',
    Influence: 'Influence'
};

export function getAxisValue(snapshot: PersonalitySnapshot, axis: PersonalityAxis): PersonalityAxisValue {
    switch (axis) {
        case 'Assertiveness':
            return {
                value: snapshot.assertiveness,
                desc: snapshot.assertiveness_description
            };
        case 'StructurePreference':
            return {
                value: snapshot.structure_preference,
                desc: snapshot.structure_preference_description
            };
        case 'Openness':
            return {
                value: snapshot.openness,
                desc: snapshot.openness_description
            };
        case 'Sociability':
            return {
                value: snapshot.sociability,
                desc: snapshot.sociability_description
            };
        case 'Influence':
            return {
                value: snapshot.influence,
                desc: snapshot.influence_description
            };
    }
}

File: src-tauri/src/integrations/snapshots/personality.rs
use serde::{Deserialize, Serialize};
use crate::sim::person::personality_matrix::{PersonalityAxis, PersonalityMatrix};

#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct PersonalitySnapshot {
    pub assertiveness: i8,
    pub structure_preference: i8,
    pub openness: i8,
    pub sociability: i8,
    pub influence: i8,
    pub assertiveness_description: String,
    pub structure_preference_description: String,
    pub openness_description: String,
    pub sociability_description: String,
    pub influence_description: String,
}

impl From<&PersonalityMatrix> for PersonalitySnapshot {
    fn from(matrix: &PersonalityMatrix) -> Self {
        PersonalitySnapshot {
            assertiveness: matrix.assertiveness,
            structure_preference: matrix.structure_preference,
            openness: matrix.openness,
            sociability: matrix.sociability,
            influence: matrix.influence,
            assertiveness_description: matrix.describe_axis(PersonalityAxis::Assertiveness),
            structure_preference_description: matrix
                .describe_axis(PersonalityAxis::StructurePreference),
            openness_description: matrix.describe_axis(PersonalityAxis::Openness),
            sociability_description: matrix.describe_axis(PersonalityAxis::Sociability),
            influence_description: matrix.describe_axis(PersonalityAxis::Influence),
        }
    }
}

impl PartialEq<&PersonalityMatrix> for PersonalitySnapshot {
    fn eq(&self, other: &&PersonalityMatrix) -> bool {
        self.assertiveness == other.assertiveness
            && self.structure_preference == other.structure_preference
            && self.openness == other.openness
            && self.sociability == other.sociability
            && self.influence == other.influence
    }
}